src/electron/routing_service.ts

```diff
@@ -279,16 +291,43 @@ async function installLinuxRoutingServices(): Promise<void> {
   const srcFolderPath = path.join(getAppPath(), OUTLINE_PROXY_CONTROLLER_PATH);
 
   console.log(`copying service installation files to ${tmp}`);
-  for (const [filename, executable] of LINUX_SERVICE_FILE_NAMES) {
-    const dest = path.join(tmp, filename);
-    await fsextra.copy(path.join(srcFolderPath, filename), dest, {overwrite: true});
-    if (executable) {
+  for (const descriptor of LINUX_SERVICE_FILE_DESCRIPTORS) {
+    const src = path.join(srcFolderPath, descriptor.filename);
+
+    const srcContent = await fsextra.readFile(src);
+    descriptor.sha256 = createHash('sha256')
+      .update(srcContent)
+      .digest('hex');
+
+    const dest = path.join(tmp, descriptor.filename);
+    await fsextra.copy(src, dest, {overwrite: true});
+
+    if (descriptor.executable) {
       await fsextra.chmod(dest, 0o755);
     }
   }
   console.log(`all service installation files copied to ${tmp} successfully`);
 
-  await executeCommandAsRoot(path.join(tmp, LINUX_INSTALLER_FILENAME));
+  // At this time, the user running Outline (who is not root) could replace these installation
+  // files in "/tmp/xxx" folder with any arbitrary scripts (because "/tmp/xxx" folder and its
+  // contents are writable by this user). Our system will then run it using root and cause a
+  // potential security breach. Therefore we need to make sure the files are the ones provided
+  // by us:
+  //   1. `chattr +i`, set the immutable flag, the flag can only be cleared by root
+  //   2. `shasum -c`, check them against our checksums embedded in AppImage
+  //   3. Run the installation script
+  //   4. `chattr -i`, always clear the immutable flag, so they can be deleted later
+  let command = `trap "/usr/bin/chattr -R -i ${tmp}" EXIT`;
+  command += `; /usr/bin/chattr -R +i ${tmp}`;
+  command +=
+    ' && ' +
+    LINUX_SERVICE_FILE_DESCRIPTORS.map(
+      ({filename, sha256}) => `/usr/bin/echo "${sha256}  ${path.join(tmp, filename)}" | /usr/bin/shasum -a 256 -c`
+    ).join(' && ');
+  command += ` && "${path.join(tmp, LINUX_INSTALLER_FILENAME)}"`;
+
+  console.log('trying to run command as root: ', command);
+  await executeCommandAsRoot(command);
 }
```

Comment:
Should we delete the files here?

Reply:
They will be deleted automatically after the reboot (of Ubuntu).